core/src/main/scala/JwtClaim.scala

```diff
@@ -19,39 +25,105 @@ case class JwtClaim(
-  def startsAt(seconds: Long): JwtClaim = this.copy(notBefore = Option(seconds))
-  def startsNow: JwtClaim = this.copy(notBefore = Option(JwtTime.nowSeconds))
+  def to(audience: Set[String]): JwtClaim = {
+    JwtClaim(content, issuer, subject, Some(audience), expiration, notBefore, issuedAt, jwtId)
+  }
 
-  def issuedIn(seconds: Long): JwtClaim = this.copy(issuedAt = Option(JwtTime.nowSeconds + seconds))
-  def issuedAt(seconds: Long): JwtClaim = this.copy(issuedAt = Option(seconds))
-  def issuedNow: JwtClaim = this.copy(issuedAt = Option(JwtTime.nowSeconds))
+  def about(subject: String): JwtClaim = {
+    JwtClaim(content, issuer, Option(subject), audience, expiration, notBefore, issuedAt, jwtId)
+  }
+
+  def withId(id: String): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, notBefore, issuedAt, Option(id))
+  }
+
+  def expiresIn(seconds: Long): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, Option(JwtTime.nowSeconds + seconds), notBefore, issuedAt, jwtId)
+  }
+
+  def expiresAt(seconds: Long): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, Option(seconds), notBefore, issuedAt, jwtId)
+  }
+
+  def expiresNow: JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, Option(JwtTime.nowSeconds), notBefore, issuedAt, jwtId)
+  }
+
+  def startsIn(seconds: Long): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, Option(JwtTime.nowSeconds + seconds), issuedAt, jwtId)
+  }
+
+  def startsAt(seconds: Long): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, Option(seconds), issuedAt, jwtId)
+  }
+
+  def startsNow: JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, Option(JwtTime.nowSeconds), issuedAt, jwtId)
+  }
+
+  def issuedIn(seconds: Long): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, notBefore, Option(JwtTime.nowSeconds + seconds), jwtId)
+  }
+
+  def issuedAt(seconds: Long): JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, notBefore, Option(seconds), jwtId)
+  }
+
+  def issuedNow: JwtClaim = {
+    JwtClaim(content, issuer, subject, audience, expiration, notBefore, Option(JwtTime.nowSeconds), jwtId)
+  }
 
-  def isValid: Boolean = JwtTime.nowIsBetweenSeconds(this.notBefore, this.expiration)
-  def isValid(issuer: String): Boolean = this.issuer.exists(_ == issuer) && this.isValid
   def isValid(issuer: String, audience: String): Boolean = this.audience.exists(_ contains audience) && this.isValid(issuer)
+
+  def isValid(issuer: String): Boolean = this.issuer.contains(issuer) && this.isValid
+
+  def isValid: Boolean = JwtTime.nowIsBetweenSeconds(this.notBefore, this.expiration)
+
+  // equality code
+  def canEqual(other: Any): Boolean = other.isInstanceOf[JwtClaim]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: JwtClaim =>
+      (that canEqual this) &&
```

Comment:
Not sure to fully understand that part...

- `this` must be a `JwtClaim` since that's the current class.
- `that` is also a `JwtClaim` from the pattern matching.

Why is there a need to call `canEqual` then?

Reply:
So basically every case class has auto-generated hashCode and equals method that allows for "==" comparison based on attribute values instead of instance references (which is default implementation from Any(?) if no equals method is provided.

A scala class does not have it. So it uses either the default impl. from Any(???) or one has to provide different implementation. I used IntelliJ Idea auto-generation function to generate these three methods (hashCode, equals and canEqual). One of the explanations on how the comparison work and why do we need that prime number (31) is here: https://alvinalexander.com/scala/how-to-define-equals-hashcode-methods-in-scala-object-equality

There were two reasons why I added these to JwtClaim:
1. "should decodeAll" spec was failing in jwt-scala\json\common\src\test\scala\JwtJsonCommonSpec.scala because of assertResult failing for claimClass if assertion is made with default equals method (since there are two different instances of claimClass):
2. because of the different json implementations I am just unsure if the equality based on attribute values is relevant or not (e.g. for some serialization stuff in one of the json implementations that I cannot comprehend since I don't know your code in detail)

So to retain original behavior (as in case class having equals method based on its attributes) I decided to generate hashCode and equals for the class aswell.